fix(forms): fix triage logic for server errors and improve field tsdoc

mmalerba opened the pull request.

mmalerba committed:
docs(forms): Improve tsdoc for field structure

Adds / improves tsdoc for field/structure.ts

mmalerba committed:
fix(forms): fix server error triage logic

fixes a bug where server errors could be lost if targeting multiple of
them at the same field

michael-small commented on packages/forms/experimental/src/api/structure.ts:
```diff
@@ -345,15 +345,34 @@ export async function submit<TValue>(
+/**
+ * Sets a list of server errors to their individual fields.
+ *
+ * @param submittedField The field that was submitted, resuling in the errors.
```
```suggestion
 * @param submittedField The field that was submitted, resulting in the errors.
```

michael-small commented on packages/forms/experimental/src/field/structure.ts:
```diff
@@ -163,16 +176,26 @@ export class ChildFieldNodeStructure extends FieldNodeStructure {
   }
 
+  /**
+   * Creates a the structure for a child field node in a field tree.
```
```suggestion
   * Creates the structure for a child field node in a field tree.
```

michael-small commented on packages/forms/experimental/src/field/structure.ts:
```diff
@@ -163,16 +176,26 @@ export class ChildFieldNodeStructure extends FieldNodeStructure {
+   * @param node The full field node that this structure belongs to
+   * @param pathNode The path corresponding to this node in the schema
+   * @param logic The logic to apply to this field
+   * @param fieldManager The field manager for this field
```
Copy/pasted from one of the other classes, needs its own params

mmalerba committed:
docs: fix typo

Co-authored-by: michael-small <33669563+michael-small@users.noreply.github.com>

mmalerba committed:
docs: fix typo

Co-authored-by: michael-small <33669563+michael-small@users.noreply.github.com>

mmalerba committed:
docs: fix typo

Co-authored-by: michael-small <33669563+michael-small@users.noreply.github.com>

mmalerba committed:
docs: address feedback

mmalerba committed:
refactor(forms): address ai review

address comments from gemini bot

alxhub commented on packages/forms/experimental/src/field/structure.ts:
```diff
@@ -36,31 +31,41 @@ export interface DataEntry {
-   * Added to array elements for tracking purposes.
-   */
+  /** Added to array elements for tracking purposes. */
+  // TODO: given that we don't ever let a field move between parents, is it safe to just extract
```
Technically I think nothing stops someone from putting the same object in multiple arrays, which is one of the reasons we do individual tracking.

leonsenft commented on packages/forms/experimental/src/field/structure.ts:
```diff
@@ -196,7 +222,9 @@ export class ChildFieldNodeStructure extends FieldNodeStructure {
         const parentValue = parent.structure.value();
         if (!isArray(parentValue)) {
           // The parent is no longer an array?
```
Is this meant to be a question or an assertion?

mmalerba replied:
Assertion I guess, I believe it was just questioning whether its possible to encounter this error (it shouldn't be)

mmalerba committed:
refactor(forms): clean up set server errors

Clean up the set server errors logic so it doesn't allocate unnecessary data structures.

Co-authored-by: Leon Senft <leonsenft@users.noreply.github.com>

mmalerba committed:
fixup! docs(forms): Improve tsdoc for field structure
